In [3]:
import duckdb
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

Matplotlib is building the font cache; this may take a moment.


In [4]:
# Connect to MotherDuck (automatically uses the MOTHERDUCK_TOKEN exported in your Fish shell)
con = duckdb.connect('md:')

Attempting to automatically open the SSO authorization page in your default browser.
Please open this link to login into your account: https://auth.motherduck.com/activate?user_code=WXXG-ZXLC


Token successfully retrieved ✅

You can display the token and store it as an environment variable to avoid having to log in again:
  PRAGMA PRINT_MD_TOKEN;


##  Bronze Layer Preview

In [5]:
# Fetch data directly into Pandas DataFrames for professional viewing
weather_df = con.execute("SELECT * FROM raw_weather ORDER BY time DESC LIMIT 15").df()
aq_df = con.execute("SELECT * FROM raw_air_quality ORDER BY time DESC LIMIT 15").df()

In [6]:
# Display the formatted tables
print("--- Weather Data ---")
display(weather_df.head())

--- Weather Data ---


,time,temperature_2m,relative_humidity_2m,apparent_temperature,precipitation,wind_speed_10m,city,latitude,longitude
0,2026-09-12T23:00,21.7,82,23.4,0.0,9.7,Istanbul,41.01384,28.94966
1,2026-09-12T23:00,25.8,45,26.4,0.0,2.2,Izmir,38.41273,27.13838
2,2026-09-12T23:00,24.3,22,22.2,0.0,0.9,Ankara,39.91987,32.85427
3,2026-09-12T23:00,17.9,86,17.9,0.5,12.0,London,51.50853,-0.12574
4,2026-09-12T23:00,21.2,91,24.1,0.0,5.3,Tokyo,35.68950,139.69171


In [7]:
print("\n--- Air Quality Data ---")
display(aq_df.head())


--- Air Quality Data ---


,time,pm10,pm2_5,nitrogen_dioxide,european_aqi,city,latitude,longitude
0,2026-09-12T23:00,13.6,8.7,22.8,37.0,Istanbul,41.01384,28.94966
1,2026-09-12T23:00,26.1,16.8,46.8,52.0,Izmir,38.41273,27.13838
2,2026-09-12T23:00,28.6,22.1,64.7,62.0,Ankara,39.91987,32.85427
3,2026-09-12T23:00,9.1,5.9,15.1,27.0,London,51.50853,-0.12574
4,2026-09-12T23:00,22.8,21.2,50.1,54.0,Tokyo,35.68950,139.69171


## Silver Layer Preview

In [8]:
stg_weather_preview = con.execute("SELECT * FROM stg_weather ORDER BY measured_at DESC LIMIT 5").df()
stg_aq_preview = con.execute("SELECT * FROM stg_air_quality ORDER BY measured_at DESC LIMIT 5").df()

In [9]:
print("--- Silver Layer: Staging Weather ---")
display(stg_weather_preview)

--- Silver Layer: Staging Weather ---


,city,measured_at,latitude,longitude,temperature_2m,relative_humidity_2m,apparent_temperature,precipitation,wind_speed_10m
0,Istanbul,2026-09-12 23:00:00,41.01384,28.94966,21.7,82.0,23.4,0.0,9.7
1,Izmir,2026-09-12 23:00:00,38.41273,27.13838,25.8,45.0,26.4,0.0,2.2
2,Ankara,2026-09-12 23:00:00,39.91987,32.85427,24.3,22.0,22.2,0.0,0.9
3,London,2026-09-12 23:00:00,51.50853,-0.12574,17.9,86.0,17.9,0.5,12.0
4,Tokyo,2026-09-12 23:00:00,35.68950,139.69171,21.2,91.0,24.1,0.0,5.3


In [10]:
print("\n--- Silver Layer: Staging Air Quality ---")
display(stg_aq_preview)


--- Silver Layer: Staging Air Quality ---


,city,measured_at,latitude,longitude,pm10,pm2_5,nitrogen_dioxide,european_aqi
0,Istanbul,2026-09-12 23:00:00,41.01384,28.94966,13.6,8.7,22.8,37.0
1,Izmir,2026-09-12 23:00:00,38.41273,27.13838,26.1,16.8,46.8,52.0
2,Ankara,2026-09-12 23:00:00,39.91987,32.85427,28.6,22.1,64.7,62.0
3,London,2026-09-12 23:00:00,51.50853,-0.12574,9.1,5.9,15.1,27.0
4,Tokyo,2026-09-12 23:00:00,35.68950,139.69171,22.8,21.2,50.1,54.0


## Gold Layer Preview

In [11]:
# Load the first 5 rows of each Gold layer table
df_fact = con.sql("SELECT * FROM my_db.main.fact_climate LIMIT 5").df()
df_dim_loc = con.sql("SELECT * FROM my_db.main.dim_location LIMIT 5").df()
df_dim_date = con.sql("SELECT * FROM my_db.main.dim_datetime LIMIT 5").df()

In [12]:
print("--- Fact Climate ---")
display(df_fact)

--- Fact Climate ---


,city,measured_at,temperature_2m,relative_humidity_2m,apparent_temperature,precipitation,wind_speed_10m,pm10,pm2_5,nitrogen_dioxide,european_aqi
0,Istanbul,2026-09-07 00:00:00,23.3,69.0,24.1,0.0,12.4,14.5,8.3,17.2,30.0
1,Istanbul,2026-09-07 01:00:00,23.2,67.0,23.8,0.0,11.2,12.8,7.7,16.4,29.0
2,Istanbul,2026-09-07 02:00:00,22.9,65.0,23.6,0.0,9.4,11.8,7.1,14.7,26.0
3,Istanbul,2026-09-07 03:00:00,22.5,64.0,23.2,0.0,7.5,11.9,7.0,14.3,26.0
4,Istanbul,2026-09-07 04:00:00,22.0,65.0,22.7,0.0,6.6,12.4,7.4,14.8,26.0


In [13]:
print("\n--- Dim Location ---")
display(df_dim_loc)


--- Dim Location ---


,city,latitude,longitude
0,London,51.50853,-0.12574
1,Madrid,40.41650,-3.70256
2,Jakarta,-6.21462,106.84513
3,Paris,48.85341,2.34880
4,Lisbon,38.72509,-9.14980


In [14]:
print("\n--- Dim Datetime ---")
display(df_dim_date)



--- Dim Datetime ---


,measured_at,date_day,year,month,day,hour
0,2020-08-26 11:00:00,2020-08-26,2020,8,26,11
1,2020-08-26 20:00:00,2020-08-26,2020,8,26,20
2,2020-10-08 06:00:00,2020-10-08,2020,10,8,6
3,2020-10-16 01:00:00,2020-10-16,2020,10,16,1
4,2020-10-26 02:00:00,2020-10-26,2020,10,26,2


## Aggregated Data

In [18]:
# Join Fact and Dimensions using DuckDB SQL
query = """
    SELECT 
        f.*,
        d.year, d.month, d.day, d.hour,
        l.latitude, l.longitude
    FROM my_db.main.fact_climate f
    LEFT JOIN my_db.main.dim_datetime d 
        ON f.measured_at = d.measured_at
    LEFT JOIN my_db.main.dim_location l 
        ON f.city = l.city
    ORDER BY f.measured_at DESC
"""

In [19]:
# Execute in DuckDB and output to a Pandas DataFrame
weather_data = con.execute(query).df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [20]:
# Verify the combined structure
display(weather_data.head())
display(weather_data.info())


,city,measured_at,temperature_2m,relative_humidity_2m,apparent_temperature,precipitation,wind_speed_10m,pm10,pm2_5,nitrogen_dioxide,european_aqi,year,month,day,hour,latitude,longitude
0,Istanbul,2026-09-12 23:00:00,21.7,82.0,23.4,0.0,9.7,13.6,8.7,22.8,37.0,2026,9,12,23,41.01384,28.94966
1,Izmir,2026-09-12 23:00:00,25.8,45.0,26.4,0.0,2.2,26.1,16.8,46.8,52.0,2026,9,12,23,38.41273,27.13838
2,Ankara,2026-09-12 23:00:00,24.3,22.0,22.2,0.0,0.9,28.6,22.1,64.7,62.0,2026,9,12,23,39.91987,32.85427
3,London,2026-09-12 23:00:00,17.9,86.0,17.9,0.5,12.0,9.1,5.9,15.1,27.0,2026,9,12,23,51.50853,-0.12574
4,Tokyo,2026-09-12 23:00:00,21.2,91.0,24.1,0.0,5.3,22.8,21.2,50.1,54.0,2026,9,12,23,35.68950,139.69171


<class 'pandas.DataFrame'>
RangeIndex: 1667928 entries, 0 to 1667927
Data columns (total 17 columns):
 #   Column                Non-Null Count    Dtype         
---  ------                --------------    -----         
 0   city                  1667928 non-null  str           
 1   measured_at           1667928 non-null  datetime64[us]
 2   temperature_2m        1667928 non-null  float64       
 3   relative_humidity_2m  1667928 non-null  float64       
 4   apparent_temperature  1667928 non-null  float64       
 5   precipitation         1667928 non-null  float64       
 6   wind_speed_10m        1667928 non-null  float64       
 7   pm10                  685734 non-null   float64       
 8   pm2_5                 685734 non-null   float64       
 9   nitrogen_dioxide      685734 non-null   float64       
 10  european_aqi          685734 non-null   float64       
 11  year                  1667928 non-null  int64         
 12  month                 1667928 non-null  int64        

None

## Data Information

**Weather & Climate Metrics**

* **temperature_2m**: The raw air temperature measured exactly 2 meters above ground level in Celsius. There is no universal "normal" here, as the baseline fluctuates drastically depending on whether you are querying a summer day in Jakarta or a winter night in Stockholm.
* **apparent_temperature**: This is the "feels like" temperature in Celsius, which directly dictates human comfort. It calculates what the body actually experiences by combining the raw temperature with humidity and wind speed. High humidity blocks sweat from evaporating (making it feel much hotter), while high wind speeds strip away body heat (making it feel colder).
* **relative_humidity_2m**: The percentage of moisture currently held in the air. A healthy, comfortable range for humans is typically between 30% and 50%. Once humidity pushes past 70%, the air starts to feel oppressive and sticky.
* **precipitation**: The total volume of rain or snow measured in millimeters. Zero is a standard dry day. As a benchmark, anything above 10mm falling in a single hour indicates a very heavy, disruptive downpour.
* **wind_speed_10m**: The speed of the wind measured 10 meters above the ground in kilometers per hour. A reading of 10 to 20 km/h is a standard light breeze. Once this number crosses 40 km/h, you are looking at strong winds that can impact daily activities.

**Air Quality & Health Metrics**

For all pollution metrics, lower numbers are always healthier.

* **pm2_5**: Fine inhalable particles that are less than 2.5 micrometers wide. This is the most dangerous metric in your dataset. Because they are so small, these particles bypass the body's natural defenses, penetrating deep into the lungs and entering the bloodstream. The World Health Organization (WHO) states the 24-hour average must stay below **15 μg/m³**. Spikes in this column indicate severe smog, heavy industrial exhaust, or nearby wildfires.
* **pm10**: Coarse particles under 10 micrometers, such as dust, pollen, and construction dirt. The WHO safety threshold is below **45 μg/m³**. While high levels will cause immediate coughing and eye irritation, they are not as deeply toxic as PM2.5.
* **nitrogen_dioxide**: A toxic gas (NO2) measured in μg/m³ that is primarily produced by burning fossil fuels. The WHO health limit is **25 μg/m³**. Tracking this column is incredibly useful for business intelligence because NO2 levels map almost perfectly to vehicle traffic and power plant usage, allowing you to visualize exact rush-hour congestion patterns across your 15 cities.
* **european_aqi**: The European Air Quality Index. Instead of making executives read raw chemical volumes, this is a master metric that aggregates all pollutants into one standardized score. A score from **0 to 50** means the air quality is Good or Fair. Once the score crosses **100**, the air is officially Poor and hazardous to sensitive groups.

**Spatial & Temporal Keys**

* **Geography**: The `city`, `latitude`, and `longitude` columns pinpoint exactly where the physical reading occurred.
* **Time Series**: The `measured_at` timestamp is broken down into `year`, `month`, `day`, and `hour` to allow for easy seasonal grouping in Power BI.

You will notice the air quality columns only have 685,734 rows, while the weather columns have 1.66 million. This is because Open-Meteo's historical air quality archive only began recording in July 2022, whereas our  weather history goes back a full 10 years.
